addons/binding/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/handler/KeContactP20Handler.java

```diff
@@ -0,0 +1,967 @@
+									theChannel, e.getMessage());
+							error = true;
+						}
+
+						if (error) {
+							logger.info(
+									"Disconnecting '{}' because of a socket error",
+									getThing().getUID().toString());
+							try {
+								theChannel.close();
+							} catch (IOException e) {
+								logger.warn(
+										"An exception occurred while closing the channel '{}': {}",
+										datagramChannel, e.getMessage());
+							}
+
+							onConnectionLost();
+
+						}
+					}
+				}
+			}
+		} finally {
+			lock.unlock();
+		}
+	}
+
+	public void onConnectionLost() {
+		if (getThing().getStatus() == ThingStatus.ONLINE) {
+			updateStatus(ThingStatus.OFFLINE);
+		}
+		establishConnection();
+	}
+
+	public void onConnectionResumed() {
+		if (getThing().getStatus() == ThingStatus.OFFLINE) {
+			updateStatus(ThingStatus.ONLINE);
+		}
+	}
+
+	private void establishConnection() {
+		lock.lock();
+		try {
+			if (getConfig().get(IP_ADDRESS) != null) {
+
+				try {
+					datagramChannel = DatagramChannel.open();
+				} catch (Exception e2) {
+					logger.error(
+							"An exception occurred while opening a datagram channel",
+							e2.getMessage());
+				}
+
+				try {
+					datagramChannel.configureBlocking(false);
+				} catch (IOException e2) {
+					logger.error(
+							"An exception occurred while configurating a datagram channel: {}",
+							e2.getMessage());
+				}
+
+				synchronized (selector) {
+					selector.wakeup();
+					int interestSet = SelectionKey.OP_READ
+							| SelectionKey.OP_WRITE;
+					try {
+						datagramChannelKey = datagramChannel.register(selector,
+								interestSet);
+					} catch (ClosedChannelException e1) {
+						logger.error(
+								"An exception occurred while registering a selector: {}",
+								e1.getMessage());
+					}
+				}
+
+				InetSocketAddress remoteAddress = new InetSocketAddress(
+						(String) getConfig().get(IP_ADDRESS),
+						REMOTE_PORT_NUMBER);
+
+				try {
+					logger.trace("Connecting the channel for {} ",
+							remoteAddress);
+					datagramChannel.connect(remoteAddress);
+
+					onConnectionResumed();
+
+				} catch (Exception e) {
+					logger.error(
+							"An exception occurred while connecting connecting to '{}:{}' : {}",
+							new Object[] {
+									(String) getConfig().get(IP_ADDRESS),
+									REMOTE_PORT_NUMBER, e.getMessage() });
+				}
+
+			}
+		} finally {
+			lock.unlock();
+		}
+	}
+
+	private Runnable connectionRunnable = new Runnable() {
+
+		@Override
+		public void run() {
+			lock.lock();
+			try {
+
+				if (datagramChannel == null || !datagramChannel.isConnected()) {
+					onConnectionLost();
+				}
+
+				if (datagramChannel.isConnected()) {
+
+					long stamp = System.currentTimeMillis();
+					if (!InetAddress.getByName(
+							((String) getConfig().get(IP_ADDRESS)))
+							.isReachable(PING_TIME_OUT)) {
+						logger.debug("Ping timed out after '{}' milliseconds",
+								System.currentTimeMillis() - stamp);
+						logger.info(
```

Comment:
remove this logger - rather keep a message that you pass into onConnectionLost() to update the status with more detail